### PR TITLE
Feature/63 store user location

### LINKED ISSUE #63: Store user's location
Create an endpoint that receives & stores the user's location 
Update openAPI code and version
Testing

## Story

#1 

## Estimate

2h

### PR DESCRIPTION
**Pull request by KseniaBel**
Enable Postgis and implement last user location #63
* fixing prod deployment

### CODE REVIEW

File: api/openapi.yaml [hunk 1]
```diff
@@ -782,18 +815,10 @@ components:
...
-          readOnly: true
         tags:
           type: array
           items:
             $ref: '#/components/schemas/Tag'
-        conversations:
-          readOnly: true
-          type: array
-          items:
-            $ref: '#/components/schemas/Conversation'
```
Comment (bschne) [inline api/openapi.yaml:796]: Why does this PR remove the `conversations` from the `User`?

File: api/openapi.yaml [hunk 2]
```diff
@@ -782,18 +815,10 @@ components:
         profilePicture:
           type: string
           readOnly: true
-        latestLocation:
-          $ref: '#/components/schemas/Coordinate'
-          readOnly: true
```
Comment (bschne) [inline api/openapi.yaml:787]: Why is this removed from the `User` here?

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java [hunk 3]
```diff
@@ -91,4 +96,10 @@ public UsersApiController(NativeWebRequest request, UserService userService, Cha
         }
     }
 
+    @Override
+    public ResponseEntity<Void> updateLocation(@ApiParam(value = "Numeric ID of the user to update",required=true) @PathVariable("userId") Long userId,@ApiParam(value = "Coordinate object that needs to be updated in user with userId" ,required=true )  @Valid @RequestBody CoordinateDto coordinateDto) throws Exception {
+        Point newLocation = new GeometryFactory().createPoint(new Coordinate(coordinateDto.getLongitude(), coordinateDto.getLatitude()));
+        userService.updateUserLocation(userId, newLocation);
+        return ResponseEntity.noContent().build();
+    }
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java:104]: This should only allow updating the location of the currently signed-in `User`, else one can manipulate the locations of other users.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/entity/Dog.java [hunk 4]
```diff
@@ -39,80 +40,4 @@
...
-    }
-
-    public Double getWeight() {
-        return weight;
-    }
-
-    public void setWeight(Double weight) {
-        this.weight = weight;
-    }
-}
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/entity/Dog.java:118]: Was this removed intentionally? 🤔

File: src/main/java/ch/uzh/ifi/hase/soprafs21/entity/User.java [hunk 5]
```diff
@@ -80,103 +80,6 @@
...
-        this.bio = bio;
-    }
-
-    public List<Dog> getListOfDogs() {
-        return listOfDogs;
-    }
-
-    public void setListOfDogs(List<Dog> listOfDogs) {
-        this.listOfDogs = listOfDogs;
-    }
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/entity/User.java:181]: Ditto — removed intentionally?

File: src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java [hunk 6]
```diff
@@ -51,17 +49,10 @@
   @JsonProperty("profilePicture")
   private String profilePicture;
 
-  @JsonProperty("latestLocation")
-  private CoordinateDto latestLocation;
-
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java:56]: Ditto — removed intentionally? If it's b/c of the rename to lastUserLocation, we should add the new one to the DTO IMO.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java [hunk 7]
```diff
@@ -125,4 +122,8 @@ public boolean isRequesterAndAuthenticatedUserTheSame(Long senderId) {
         return authenticatedUser.equals(sender);
     }
 
+    @Transactional
+    public void updateUserLocation(Long userId, Point newLocation) {
+        this.userRepository.updateUserLocation(userId, newLocation);
+    }
```
Comment (bschne) [inline src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java:128]: ...or rather, put the authorization I mentioned in the controller here?

**Review by bschne — CHANGES_REQUESTED** [review]

File: api/openapi.yaml (hunk 2 above)
Comment (KseniaBel) [reply api/openapi.yaml:787]: I created the new endpoint to update the user latest location, this objects will be used only for the user profile functionality.

File: api/openapi.yaml (hunk 1 above)
Comment (KseniaBel) [reply api/openapi.yaml:796]: The same reason as above. The is a separate endpoint for Conversation functionality.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java (hunk 3 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/controller/UsersApiController.java:104]: True. This will be added later(as well as other checks). 
I decided to go for the implementation of a simple functionality first(as it stated in the Milestone requirement). Later on I will add and test other flows.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/entity/Dog.java (hunk 4 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/entity/Dog.java:118]: Yes, there is Lombok annotation @Data for this.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/entity/User.java (hunk 5 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/entity/User.java:181]: The same as above.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java (hunk 6 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java:56]: Generated code. It was removed from API definition that's why it is removed from here.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java (hunk 7 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/service/UserService.java:128]: It will be added here when the security is ready.

File: api/openapi.yaml (hunk 2 above)
Comment (bschne) [reply api/openapi.yaml:787]: The issue is that we need the latestLocation when we retrieve lists of users to show them on the map & their profiles (map: obvious; profile: we show the distance between the current user and the user whose profile they are viewing), which I think uses the same `User` schema from the OpenAPI spec.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java (hunk 6 above)
Comment (bschne) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java:56]: See above issue re. removing this from User/UserDTO

File: api/openapi.yaml (hunk 2 above)
Comment (KseniaBel) [reply api/openapi.yaml:787]: The front-end is the one providing with the current user location so I assume you should take it directly from the device.
Regarding the rest of the users, we have a separate schema which is light-weight and we will include the attribute there. which  is my next task.

File: src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java (hunk 6 above)
Comment (KseniaBel) [reply src/main/java/ch/uzh/ifi/hase/soprafs21/rest/dto/UserDto.java:56]: the same reply as the previous

File: api/openapi.yaml (hunk 2 above)
Comment (bschne) [reply api/openapi.yaml:787]: The current location of the user using the app is taken from the device, correct. But the endpoints `GET /users` (also used with the radius/area filter) and `GET /user/:id` use this schema and need to have the location. The former can be switched to the lightweight schema since we only need OnlineStatus and Location to display the users on the map, and additionally the Name and ProfileImage to show them in lists. But the latter is also used to show the profile (from the map and the full profile) and we also need the dogs and tags there, so they shouldn't be moved to the lightweight schema IMO.

Comment (KseniaBel) [reply api/openapi.yaml:787]: You lost me here...I do not see why we need a whole user object to be return when we call 'GET /users' and also why we need to have a location there.
('GET /user/:id') If you refer to the screen 'User overview' when a particular user is chosen, the functionality is not implemented yet. It is a separate task.

Comment (KseniaBel) [reply api/openapi.yaml:787]: I will add the coordinate to the User object when it will be required, if you don't mind.

**Review by bschne — APPROVED** [review]
